Adding 'Is Active' check box to regional partners edit, view and index pages

vijayamanohararaj opened the pull request:
- Rails migration to backfill the new "is_active" column for regional partners to true.
 - Adding validation in model to make it a required field
 - UI changes to show the new field in edit, show and index pages
 - Fixed a broken link to show in the index page

Edit page updates with new "Is active" field
![image](https://user-images.githubusercontent.com/124813947/225384281-28dfa431-d3d6-4c10-aef9-d444cd9f57cd.png)

List page updates with new "Is active" field
![image](https://user-images.githubusercontent.com/124813947/225384853-95210b88-c9b7-48eb-a535-b55fb3339d64.png)

View page updates with new "Is active" field
![image](https://user-images.githubusercontent.com/124813947/225384924-ded74054-5c8d-4ca3-a988-d28cdafeaee1.png)

## Links

JIRA ticket: https://codedotorg.atlassian.net/browse/ACQ-251?atlOrigin=eyJpIjoiNTE3NWMzMzM1YWUwNDEyOGIzZWQ5NzM2YzU0NGU3MmMiLCJwIjoiaiJ9

Change that added the migration to include the new field in DB: https://github.com/code-dot-org/code-dot-org/pull/50680 

## Testing story

Includes tests that cover controller and model changes.

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

This change involves a back fill for all regional partners. The total number of regional partners is about 100 in production and less than 10 in staging. Hence, this shouldn't be a long blocking operation.

## Follow-up work

None

## Privacy

No

## Security

N/A

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [ ] New features are translatable or updates will not break translations
  **A new field was added in the UI, but the page is mainly for internal/partner use and doesn't have any translations at this time**
- [ ] Relevant documentation has been added or updated
  **N/A**
- [x] User impact is well-understood and desirable
- [x] Pull Request is labeled appropriately
- [x] Follow-up work items (including potential tech debt) are tracked and linked

vijayamanohararaj commented on dashboard/app/views/regional_partners/show.html.haml:
```diff
@@ -20,6 +20,12 @@
+  %strong Active:
+  - if @regional_partner.is_active
+    = 'Yes'
+  -else
```
nit:  space before "else" for consistency

vijayamanohararaj commented:
@dmcavoy , can you please take a look at the screenshots and confirm if these are inline with the requirements.

vijayamanohararaj commented on dashboard/db/schema.rb:
```diff
@@ -10,7 +10,7 @@
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2023_03_09_041659) do
+ActiveRecord::Schema.define(version: 2023_03_13_222451) do
```
@davidsbailey , how are conflicts between migrations resolved when the migration was generated much earlier than the time at which change is merged. Especially, if there are newer migrations checked in since the engineer forked staging branch.

davidsbailey replied:
good question. :-) I usually resolve the conflict arbitrarily, and then rerun `rake db:migrate` and go with whatever it generates.

vijayamanohararaj replied:
I'm getting the same sense reading [this article](https://natemacinnes.github.io/how-to-handle-schema-conflicts-when-rebasing-git-rails.html), that running rake db:migrate will generate the most recent schema.rb with the appropriate version consolidating the conflicting migrations. I guess this should work in most cases, unless the conflicting migration and the one in feature branch do a DB operation itself which has potential conflicts. I'm assuming that's a fairly rare scenario.

davidsbailey commented on dashboard/app/controllers/test_controller.rb:
```diff
@@ -235,7 +235,7 @@ def teacher_form_data
   def create_teacher_application
     return unless (user = current_user)
-    regional_partner = RegionalPartner.create!(name: "regional-partner#{Time.now.to_i}-#{rand(1_000_000)}")
+    regional_partner = RegionalPartner.create!(name: "regional-partner#{Time.now.to_i}-#{rand(1_000_000)}", is_active: true)
```
nice find!

davidsbailey commented on dashboard/app/models/regional_partner.rb:
```diff
@@ -139,6 +139,7 @@ def upcoming_summer_workshops
   validates_inclusion_of :applications_decision_emails, in: APPLICATION_DECISION_EMAILS, if: -> {applications_decision_emails.present?}
   validates :csd_cost, numericality: {greater_than: 0}, if: -> {csd_cost.present?}
   validates :csp_cost, numericality: {greater_than: 0}, if: -> {csp_cost.present?}
+  validates :is_active, inclusion: [true, false]
```
this seems a bit redundant for a boolean field. are you just trying to enforce that the field is not `nil`? if so, you could do it with `validates_presence_of :is_active`

vijayamanohararaj replied:
Yes, thanks for the suggestion, this is much better to read.

davidsbailey commented on dashboard/db/migrate/20230313222451_backfill_regional_partner_is_active.rb:
```diff
@@ -0,0 +1,6 @@
+  def change
+    RegionalPartner.reset_column_information
+    RegionalPartner.update_all(is_active: true)
+  end
```
this looks like something you only want to do if you are running the migration forward, but not if you are rolling it back, correct? if so, this method should be be renamed to `up`, and there should also be an empty `down` method.

vijayamanohararaj replied:
Yes, that is correct.

I was actually curious about that and especially on using up/down vs change. I [learnt that](https://www.gapintelligence.com/blog/up-and-down-a-rails-migration/) in more recent versions of rails, it is can deduce the up/down from the change operation and this was a preferable option to use.

I also tested this in both migrate and roll back operations and found that roll back was a no-op. I like the suggestion of using up/down for overall readability and being very explicit about the roll back.

davidsbailey commented on dashboard/db/migrate/20230313222451_backfill_regional_partner_is_active.rb:
in general, our best practice is to put migrations in separate PRs, so that if the code change needs to be reverted, we do not end up having to try to back out the migration. however, in this case it seems ok to leave it how you have it, because there is nothing to undo if this migration needs to be backed out.

vijayamanohararaj replied:
I agree in cases where the migration has a schema change. In this case particularly I wanted to include the other changes to avoid any new records getting created with nil data for the new field after the migration and before the rest of the changes are in.

davidsbailey replied:
>> in general, our best practice is to put migrations in separate PRs, so that if the code change needs to be reverted, we do not end up having to try to back out the migration. however, in this case it seems ok to leave it how you have it, because there is nothing to undo if this migration needs to be backed out.

> I agree in cases where the migration has a schema change. 

Good point, this does seem to satisfy the concern I raised above. 

> In this case particularly I wanted to include the other changes to avoid any new records getting created with nil data for the new field after the migration and before the rest of the changes are in.

I think this is probably good enough because we don't create regional partners very often, but it seems worth pointing out that this does leave room for a race condition because we [migrate dashboard db](https://codedotorg.slack.com/archives/C03CK8FGX/p1676665321454779) about 30 minutes before we [finish deploying code changes](https://codedotorg.slack.com/archives/C03CK8FGX/p1676667187170179) in production.

based on the low chances of this causing an issue, I think it's fine to ship this PR as-is. this could be a good time to talk about the best practice I mentioned though. I don't think it is documented anywhere, so most engineers hear about it in code review comments when they make a first migration PR. questions:
1. do you agree with the best practice "migrations should generally be in separate PRs from code changes. if combined, PR should include justification for why this is safe"
2. if yes, I'd love to discuss how this could be better documented (lint rule, PR template, onboarding doc, style guide, etc)

vijayamanohararaj replied:
I would like to have an overall discussion on best practices under a few different scenarios, 

1. Change include a migration with a schema change - this should be a separate PR, I agree
2. Change includes a migration with backfill or some sort of data updates - I'd like us to discuss this a bit more mainly around how to slop bleeding in case we split the PRs 

Let's try to find time in a parking lot of sprint planning one of these days for a first round of discussion.

megcrenshaw replied:
+1 for this –– I'd also love some documentation on this because I feel like different discussions detail slightly different best practices

davidsbailey replied:
sounds good, let's discuss. I think there are a few standard patterns here but there are some nuances too.

davidsbailey reviewed (approved):
nice work @vijayamanohararaj ! I am curious, do you plan to follow up with a migration to make the new column non-null? the validations provided by rails can be circumvented, but DB constraints like "not null" can be relied upon to never be violated.

I'm approving contingent upon addressing a few minor comments below. if there's anything you would like me to take another look at before you merge, please feel free to request another review from me.

dmcavoy commented:
> @dmcavoy , can you please take a look at the screenshots and confirm if these are inline with the requirements.

Changes looks great! Thanks for taking the extra time to update the list and view pages!

vijayamanohararaj commented:
> nice work @vijayamanohararaj ! I am curious, do you plan to follow up with a migration to make the new column non-null? the validations provided by rails can be circumvented, but DB constraints like "not null" can be relied upon to never be violated.
> 
> I'm approving contingent upon addressing a few minor comments below. if there's anything you would like me to take another look at before you merge, please feel free to request another review from me.

Thank you so much!! That's a good call out. Yes, I can follow up with another PR to add a DB level constraint before closing out the JIRA task.

Ack. on the comments.

@davidsbailey , thanks for the review. I have addressed most of the comments, have couple open that I'd like to get your response on to close on discussion. Re-requested review for the same.

davidsbailey reviewed (approved):
looks great @vijayamanohararaj ! I think I responded to all the open topics of conversation, but please let me know if I missed one.

megcrenshaw commented:
Thanks for the detailed discussions, all! I learned a lot from reading this
